yew-router/src/service.rs

```diff
@@ -146,53 +147,52 @@ where
-            log::error!("Could not serialize state string");
-            "".to_string()
+        let state_buf = bincode::serialize(&state).unwrap_or_else(|_| {
+            log::error!("Could not serialize state");
+            Vec::default()
         });
         cfg_match! {
             feature = "std_web" => ({
-                self.history.push_state(state_string, "", Some(route));
+                self.history.push_state(TypedArray::from(state_buf.as_slice()), "", Some(route));
             }),
             feature = "web_sys" => ({
-                let _ = self.history.push_state_with_url(&Value::from_str(&state_string), "", Some(route));
+                let _ = self.history.push_state_with_url(&Uint8Array::from(state_buf.as_slice()).into(), "", Some(route));
             }),
         };
     }
 
     /// Replaces the route with another one removing the most recent history event and
     /// creating another history event in its place.
     pub fn replace_route(&mut self, route: &str, state: STATE) {
-        let state_string: String = serde_json::to_string(&state).unwrap_or_else(|_| {
-            log::error!("Could not serialize state string");
-            "".to_string()
+        let state_buf = bincode::serialize(&state).unwrap_or_else(|_| {
+            log::error!("Could not serialize state");
+            Vec::default()
         });
         cfg_match! {
             feature = "std_web" => ({
-                let _ = self.history.replace_state(state_string, "", Some(route));
+                let _ = self.history.replace_state(TypedArray::from(state_buf.as_slice()), "", Some(route));
             }),
             feature = "web_sys" => ({
-                let _ = self.history.replace_state_with_url(&Value::from_str(&state_string), "", Some(route));
+                let _ = self.history.replace_state_with_url(&Uint8Array::from(state_buf.as_slice()).into(), "", Some(route));
             }),
         };
     }
 
     /// Gets the concatenated path, query, and fragment.
     pub fn get_route(&self) -> Route<STATE> {
         let route_string = Self::get_route_from_location(&self.location);
-        let state: STATE = get_state_string(&self.history)
+        let state: STATE = get_state_buf(&self.history)
             .or_else(|| {
                 log::trace!("History state is empty");
                 None
             })
-            .and_then(|state_string| -> Option<STATE> {
-                serde_json::from_str(&state_string)
+            .and_then(|state_buf| -> Option<STATE> {
+                bincode::deserialize(&state_buf)
                     .ok()
                     .or_else(|| {
-                        log::error!("Could not deserialize state string");
+                        log::error!("Could not deserialize state");
                         None
                     })
-                    .and_then(std::convert::identity) // flatten
```

Comment:
This changed the deserialized type to `Option<STATE>` instead of `STATE`. That worked with serde_json, because it flattens `Option<STATE>` to `null | STATE` (i.e. as opposed to `null | { some: STATE }`, but it does not work with bincode, which does not do similar flattening.